packages/styletron-engine-monolithic/src/client/client.js

```diff
@@ -0,0 +1,142 @@
+// @flow
+
+/* eslint-env browser */
+
+import {StyleSheet} from "./sheet";
+
+import type {
+  StandardEngine,
+  KeyframesObject,
+  FontFaceObject,
+  StyleObject,
+} from "styletron-standard";
+
+import injectStylePrefixed from "../inject-style-prefixed.js";
+
+import {
+  keyframesBlockToRule,
+  declarationsToBlock,
+  keyframesToBlock,
+  fontFaceBlockToRule,
+  hashCssObject,
+} from "../css.js";
+
+declare var __DEV__: boolean;
+
+type hydrateT =
+  | HTMLCollection<HTMLStyleElement>
+  | Array<HTMLStyleElement>
+  | NodeList<HTMLStyleElement>;
+
+type optionsT = {
+  hydrate?: hydrateT,
+  container?: Element,
+  prefix?: string,
+};
+
+type cacheT = {
+  [key: string]: boolean,
+};
+
+class StyletronClient implements StandardEngine {
+  container: Element;
+  opts: optionsT;
+  fontFaceSheet: HTMLStyleElement;
+  keyframesSheet: HTMLStyleElement;
+
+  styleSheet: StyleSheet;
+  styleCache: cacheT;
+  keyframesCache: cacheT;
+  fontFaceCache: cacheT;
+
+  constructor(opts?: optionsT = {}) {
+    this.opts = opts || {};
+    this.styleCache = {};
+    this.fontFaceCache = {};
+    this.keyframesCache = {};
+
+    if (opts.container) {
+      this.container = opts.container;
+    }
+
+    // hydration
+    if (opts.hydrate && opts.hydrate.length > 1) {
+      if (__DEV__) {
+        // eslint-disable-next-line no-console
+        console.warn(
+          "Client hydration expects only a single style tag from the server. Make sure your server uses styletron-engine-monolithic and not atomic.",
+        );
+      }
+    } else if (opts.hydrate && opts.hydrate.length === 1) {
+      // infer container from parent element
+      if (!this.container) {
+        const parentElement = opts.hydrate[0].parentElement;
+        if (parentElement !== null && parentElement !== void 0) {
+          this.container = parentElement;
+        }
+      }
+      // there is a single style tag coming from the monolithic server
+      const element = opts.hydrate[0];
+      const dataHydrate = element.getAttribute("data-hydrate");
+      if (dataHydrate) {
+        dataHydrate.split(" ").forEach(hashKey => {
+          // cache keys are unique across fonts, keyframes and other css so
+          // we use a single attribute to get them all
+          this.styleCache[hashKey] = true;
+          this.fontFaceCache[hashKey] = true;
+          this.keyframesCache[hashKey] = true;
```

Comment:
If we hydrate in this manner, I wonder if it'd be preferable to just share the same cache reference for all three aliases.

Since I think the only benefit of separate caches would be smaller size and thus faster lookups, but if they contain the same keys this benefit seems greatly reduced.